### ISSUE TITLE
Either seems buggy

### ISSUE THREAD

**Issue by mvh77** [javaslang/javaslang#881]
```java
Either<String, String> good = Either.right("good");
Either<String, String> bad = Either.right("bad");

Either<String, Tuple2<String, String>> result = good.right().flatMap(g -> bad.right().map(b -> Tuple.of(g,b))).toEither();
System.out.println(result); // prints Right((good, bad))
```
It isn't right! ;) It's left!

Scala workbench:
```scala
val good: Either[String,String] = Right("good")   //> good  : Either[String,String] = Right(good)
val bad: Either[String,String]  = Left("bad")     //> bad  : Either[String,String] = Left(bad)
good.right.flatMap { g => bad.right.map { b => (g,b) } }
                                                  //> res4: scala.util.Either[String,(String, String)] = Left(bad)
```

**Comment by danieldietrich** [comment]
Nice catch, thx!

**Comment by danieldietrich** [comment]
Is it right that you used right instead of left?

Yours:

```java
Either<String, String> bad = Either.right("bad");
```

But it should be like the Scala example:

```java
Either<String, String> bad = Either.left("bad");
```

**Comment by danieldietrich** [comment]
Instead we could also use:

```java
Either<String, String> good = new Right<>("good")
Either<String, String> bad = new Left<>("bad");
```

**Comment by mvh77** [comment]
No, it should be left of course!  My bad! But that gives you the original problem I had, which is an exception!

Exception in thread "main" java.util.NoSuchElementException: Either.right().get() on Left

**Comment by danieldietrich** [comment]
That's strange. It should not happen, Either once was unit-tested very well. I will take a look later today.

Additionally I filed an issue to behave more like Scala: #882

**Comment by mvh77** [comment]
On your other comment above, would it make sense to have also static constructors on implementation classes, so you'd have Right.of(), Left.of(), Some.of(), Tuple2.of() to be in line with the 'static' contructors  of Scala. I know there are static constructors on the base type (like Option.of) already. Sometimes these behave differently in Scala though (Option(null) vs Some(null)).

**Comment by danieldietrich** [comment]
We currently use static factory methods on interface methods only.

Additionally implementing classes have exactly one constructor in Javaslang. If possible, this constructor takes exactly the arguments to initialize the final fields of the class. Like case classes of Scala.

Classes may have factory methods, if other initialization possibilities are needed. Like companion objects in Scala.

I use new `new Some<>(a)` instead of `Option.of(a)`.

---

Most probably there will no `Xxx.of` methods for impl classes. I once had the idea to provide factory methods named like the class, to offer more Scala-like API. 

Examples:

```java
interface Option<T> {

    None None = None.instance();

    static <T> Option<T> Option(T t) {
        return (t == null) ? None : Some(t);
    }

    static <T> Some<T> Some(T t) {
        return new Some<>(t);
    }
```

Usage:

```java
import static javaslang.control.Option.*;

class API {{

    Option<String> option = Some("hi");

}}
```

But this is a hack. I want Javaslang behave like users expect standard Java to work.

**Comment by danieldietrich** [comment]
Btw - I like your new profile pic! Asked myself who you are before you added your name :-)

**Comment by mvh77** [comment]
Hmm.. but we have static init methods on all collections, so this seems like a discrepency to me. I can say Vector.of("foo") but not Some.of("foo"). A static ".of()"  seems to me like Scala's apply on an companion Object, so Some("foo") in Scala would translate to Some.of("foo") in Java. Anyway, my .02$ as usual. :smile:

**Comment by danieldietrich** [comment]
I think you're right. There should be no inherent - undocumented logic or some conventions which probably no one knows and which most probably eventually will be broken for that reason.

We should file an issue and do it.

**Comment by danieldietrich** [comment]
I've added `Option.some`, see #888

**Comment by danieldietrich** [comment]
Note: The `*.of(T)` constructors are implemented in #889
